ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/Web3jClientBuilder.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.ethereum.executionlayer.client;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+import okhttp3.OkHttpClient;
+import org.web3j.protocol.Web3jService;
+import org.web3j.protocol.http.HttpService;
+import org.web3j.protocol.websocket.WebSocketClient;
+import org.web3j.protocol.websocket.WebSocketService;
+import tech.pegasys.teku.ethereum.executionlayer.client.auth.JwtConfig;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
+import tech.pegasys.teku.infrastructure.time.TimeProvider;
+
+public class Web3jClientBuilder {
+  private TimeProvider timeProvider;
+  private URI endpoint;
+  private Optional<JwtConfig> jwtConfigOpt = Optional.empty();
+
+  public Web3jClientBuilder endpoint(String endpoint) {
+    this.endpoint = parseEndpoint(endpoint);
+    return this;
+  }
+
+  private URI parseEndpoint(final String endpoint) {
+    final URI endpointUri;
+    try {
+      endpointUri = new URI(endpoint);
+    } catch (URISyntaxException ex) {
+      throw new InvalidConfigurationException(
+          String.format("%s is not a correct endpoint URI", endpoint), ex);
+    }
+    return endpointUri;
+  }
+
+  public Web3jClientBuilder jwtConfigOpt(Optional<JwtConfig> jwtConfig) {
+    this.jwtConfigOpt = jwtConfig;
+    return this;
+  }
+
+  public Web3jClientBuilder timeProvider(TimeProvider timeProvider) {
+    this.timeProvider = timeProvider;
+    return this;
+  }
+
+  public Web3JClient build() {
+    checkNotNull(timeProvider);
+    checkNotNull(endpoint);
+    switch (endpoint.getScheme()) {
+      case "http":
```

Comment:
We can also support `https` using the http service.

Reply:
Maybe I'm wrong in my understanding but when you use https in browser the connection is accepted because certificates are signed by limited list of trusted companies and browser trust this signature. In case with your own TLS secured gateway you need to provide certificates keys etc. In Besu there are flags for it: 

---

Specifies the keystore using the [--rpc-http-tls-keystore-file](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-tls-keystore-file) option.
Specifies the file that contains the password to decrypt the keystore using the [--rpc-http-tls-keystore-password-file](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-tls-keystore-password-file) option.
[Specifies the clients](https://besu.hyperledger.org/en/stable/HowTo/Configure/TLS/Configure-TLS/#create-the-known-clients-file) allowed to connect to Besu using the [--rpc-http-tls-known-clients-file](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-tls-known-clients-file) option.
specifies the Java cipher suites using the [--rpc-http-tls-cipher-suite](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-tls-cipher-suite) option.
specifies the TLS protocol version using the [--rpc-http-tls-protocol](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-tls-protocol) option.

---
https://besu.hyperledger.org/en/stable/HowTo/Configure/TLS/Configure-TLS/

So to handle TLS we need to pass something like this to the client, otherwise you cannot trust this connection. So until we setup OkHttpClient with our tls keys, we cannot support it.
This way we need to implement something like this:
https://stackoverflow.com/a/67263724
or with okhttp-tls it looks less bulky:
https://github.com/square/okhttp/tree/master/okhttp-tls
Overall, we cannot trust server just because it turned https on, that's my understanding. That's why I see it as different feature

Reply:
Do we have any not self signed https api to test with? Infura or smth like this?

Reply:
There's a few ways to deal with self-signed certs - one as you point out is to build in functionality to add to the list of trusted certs.
But you can also either get a cert from somewhere that is trusted or add to the java trusted certs (https://docs.oracle.com/cd/E19906-01/820-4916/geygn/index.html).

Mostly I'd expect people not to use SSL except in big setups where they'll likely have an internal cert authority they already add as trusted for various things and can do that.

We already support https for the eth1-endpoint as it's required for use with Infura and that's the same web3j client.

Reply:
ok, right
we could definitely support it for signed certs